HADOOP-13327 Output Stream Specification.

steveloughran opened the pull request:
Specification of OutputStream and Syncable

with
* RawLocalFileSystem to implement Syncable
* Consistent use of StreamCapabilities everywhere

This is a rebase of #2102. Because the RawLocalOutputStream is now wrapped by `BufferedIOStatisticsOutputStream`, which does passthrough of stream capabilities and the Syncable API, the tests which were failing there should now work

steveloughran committed:
HADOOP-13327 Output Stream Specification.

This PR removes the changes related to S3A output stream lifecycle,
so only covers the specification of Syncable and ensures that StreamCapabilities
passes all the way through to the final implementation classes.

All streams which implement Syncable hsync/hflush declare this in their stream capabilities

Change-Id: I82b16a8e0965f34eb0c42504da43e8fbeabcb68c

steveloughran commented:
* does not address final comments in that review
* no ITest runs (I don't think it needs it; will look at again)

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 32s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |   |   0m  0s | [test4tests](test4tests) |  The patch appears to include 9 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  13m 48s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  23m 12s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  20m  4s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04  |
| +1 :green_heart: |  compile  |  17m 15s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| +1 :green_heart: |  checkstyle  |   2m 44s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   5m 43s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  24m 27s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   4m  3s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04  |
| +1 :green_heart: |  javadoc  |   4m 57s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| +0 :ok: |  spotbugs  |   0m 47s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   9m 47s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 26s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m 31s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  26m 52s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04  |
| -1 :x: |  javac  |  26m 52s | [/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt) |  root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 generated 12 new + 2038 unchanged - 0 fixed = 2050 total (was 2038)  |
| +1 :green_heart: |  compile  |  24m  4s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| -1 :x: |  javac  |  24m  4s | [/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01.txt) |  root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 generated 12 new + 1931 unchanged - 0 fixed = 1943 total (was 1931)  |
| -0 :warning: |  checkstyle  |   3m 28s | [/diff-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/diff-checkstyle-root.txt) |  root: The patch generated 4 new + 128 unchanged - 7 fixed = 132 total (was 135)  |
| +1 :green_heart: |  mvnsite  |   6m  6s |  |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s | [/whitespace-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/whitespace-eol.txt) |  The patch has 4 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  5s |  |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  19m 15s |  |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   1m 10s | [/diff-javadoc-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/diff-javadoc-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt) |  hadoop-common-project_hadoop-common-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 generated 1 new + 98 unchanged - 1 fixed = 99 total (was 99)  |
| +1 :green_heart: |  javadoc  |   5m 33s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| +1 :green_heart: |  findbugs  |  13m 45s |  |  the patch passed  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  11m 18s | [/patch-unit-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt) |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 52s |  |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  | 113m 32s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 53s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 22s |  |  hadoop-azure-datalake in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  7s |  |  The patch does not generate ASF License warnings.  |
|  |   | 364m 34s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.fs.contract.localfs.TestLocalFSContractCreate |
|   | hadoop.hdfs.server.namenode.TestReconstructStripedBlocks |
|   | hadoop.hdfs.server.namenode.TestHDFSConcat |
|   | hadoop.hdfs.server.namenode.ha.TestStandbyCheckpoints |
|   | hadoop.hdfs.TestMultipleNNPortQOP |
|   | hadoop.hdfs.server.namenode.TestFSEditLogLoader |
|   | hadoop.hdfs.server.namenode.TestAddStripedBlocks |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2587 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint xml |
| uname | Linux 1dfb2edeba04 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 2825d060cf9 |
| Default Java | Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/testReport/ |
| Max. process+thread count | 3841 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-13327. ChecksumFS streams must fail probes for syncable capabilities...

...even if the inner stream says "yes"

Change-Id: Ie401232a23fbc05ae40baf1700fe1cf2ab80a42a

steveloughran committed:
HADOOP-13327. LocalFS with write checksums disabled.

Verify that if you call LocalFileSystem.setWriteChecksum(false)
then the streams you get back support Syncable.

This now allows applications to sync to the local FS by calling
getLocal(), disabling checksums and then creating files.

+ docs reviewed, hdfs issued moved to one place, and
"why implementors must not relay flush to hflush".

* abfs.xml declares support for hflush/hsync

Change-Id: I6f1b19a00877a26a1f3a53ce15ca2a1cff205ded

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  26m  6s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |   |   0m  0s | [test4tests](test4tests) |  The patch appears to include 9 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  15m 33s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m  0s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  20m 18s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04  |
| +1 :green_heart: |  compile  |  17m 37s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| +1 :green_heart: |  checkstyle  |   2m 45s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   5m 50s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  24m 25s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   4m 28s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04  |
| +1 :green_heart: |  javadoc  |   5m 28s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| +0 :ok: |  spotbugs  |   0m 53s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  10m  6s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 27s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 50s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  21m 44s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04  |
| -1 :x: |  javac  |  21m 44s | [/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt) |  root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 generated 12 new + 2037 unchanged - 0 fixed = 2049 total (was 2037)  |
| +1 :green_heart: |  compile  |  18m 41s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| -1 :x: |  javac  |  18m 41s | [/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01.txt) |  root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 generated 12 new + 1930 unchanged - 0 fixed = 1942 total (was 1930)  |
| -0 :warning: |  checkstyle  |   2m 39s | [/diff-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/diff-checkstyle-root.txt) |  root: The patch generated 4 new + 182 unchanged - 7 fixed = 186 total (was 189)  |
| +1 :green_heart: |  mvnsite  |   5m 44s |  |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s | [/whitespace-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/whitespace-eol.txt) |  The patch has 4 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  6s |  |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  15m 35s |  |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   1m  3s | [/diff-javadoc-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/diff-javadoc-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04.txt) |  hadoop-common-project_hadoop-common-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 generated 1 new + 98 unchanged - 1 fixed = 99 total (was 99)  |
| +1 :green_heart: |  javadoc  |   5m 25s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01  |
| +1 :green_heart: |  findbugs  |  10m 53s |  |  the patch passed  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   9m 53s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 40s |  |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  | 104m 39s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 51s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 22s |  |  hadoop-azure-datalake in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 57s |  |  The patch does not generate ASF License warnings.  |
|  |   | 361m 44s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.balancer.TestBalancer |
|   | hadoop.hdfs.server.namenode.TestFsck |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.hdfs.server.datanode.TestDataNodeErasureCodingMetrics |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2587 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint xml |
| uname | Linux 08e86d292978 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 06fef5ee43d |
| Default Java | Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.18.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_275-8u275-b01-0ubuntu1~18.04-b01 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/testReport/ |
| Max. process+thread count | 3953 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

joshelser commented:
Tagging a few HBase folks who may be interested in this as well, @ndimiduk @saintstack @Apache9 @busbey given the previous work on apache/hbase#1408 and apache/hbase#1597.

Steve had mentioned to me that he was thinking of HBase and us being able to more safely put WALs onto file:// with these changes (not having to just disable the checks with LocalFileSystem).

joshelser reviewed (approved):
All makes sense to me. Tried my best to read (and not just skim) outputstream.md. Excellent work. Thanks for bringing to my attention, Steve!

steveloughran commented:
> Steve had mentioned to me that he was thinking of HBase and us being able to more safely put WALs onto file:// with these changes (not having to just disable the checks with LocalFileSystem).

you are still going to need some robust storage, RAID-1+ or similar, but we are getting sync all the way through, and you can query the streams to make sure they say they support it -including local fs.

steveloughran commented:
checkstyle nits
```
./hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCreateTest.java:512:          if (c == -1) {:24: Must have at least one statement. [EmptyBlock]
./hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java:70:import org.apache.hadoop.util.StringUtils;:8: Unused import - org.apache.hadoop.util.StringUtils. [UnusedImports]
./hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/BlockBlobAppendStream.java:30:import java.util.Locale;:8: Unused import - java.util.Locale. [UnusedImports]
./hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/SyncableDataOutputStream.java:31:import org.apache.hadoop.classification.InterfaceAudience;:1: Duplicate import to line 28 - org.apache.hadoop.classification.InterfaceAudience. [RedundantImport]
```

steveloughran committed:
HADOOP-13327: checkstyle nits

Change-Id: I3f9fa4c277d3c6975856274b898d56d1103bd742

joshelser commented:
> you are still going to need some robust storage, RAID-1+ or similar, but we are getting sync all the way through, and you can query the streams to make sure they say they support it -including local fs.

Yep, for sure. This is a nice improvement.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 41s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |   |   0m  0s | [test4tests](test4tests) |  The patch appears to include 11 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  13m 57s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  25m 26s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  27m  5s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |  23m 20s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  checkstyle  |   4m 59s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   6m 12s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  26m 34s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   4m 18s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   5m 16s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +0 :ok: |  spotbugs  |   0m 53s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  10m 16s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 39s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  19m 51s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| -1 :x: |  javac  |  19m 52s | [/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/artifact/out/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04.txt) |  root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 generated 11 new + 2037 unchanged - 0 fixed = 2048 total (was 2037)  |
| +1 :green_heart: |  compile  |  17m 51s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| -1 :x: |  javac  |  17m 51s | [/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/artifact/out/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01.txt) |  root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 generated 11 new + 1930 unchanged - 0 fixed = 1941 total (was 1930)  |
| -0 :warning: |  checkstyle  |   3m 48s | [/diff-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/artifact/out/diff-checkstyle-root.txt) |  root: The patch generated 2 new + 182 unchanged - 7 fixed = 184 total (was 189)  |
| +1 :green_heart: |  mvnsite  |   5m 45s |  |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s | [/whitespace-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/artifact/out/whitespace-eol.txt) |  The patch has 4 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  7s |  |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 50s |  |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   4m 28s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   5m 29s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  findbugs  |  10m 49s |  |  the patch passed  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m 14s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 38s |  |  hadoop-hdfs-client in the patch passed.  |
| +1 :green_heart: |  unit  | 190m 39s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 39s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 24s |  |  hadoop-azure-datalake in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  6s |  |  The patch does not generate ASF License warnings.  |
|  |   | 444m 38s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2587 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint xml |
| uname | Linux 42c98f9f9a9c 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 2a38ed0e0c0 |
| Default Java | Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/testReport/ |
| Max. process+thread count | 3591 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/3/console |
| versions | git=2.25.1 maven=3.6.3 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-13327. checkstyle and whitespace

Change-Id: I2f2fe20ab317b8836f0ba460e8d2ab95762a5a3b

steveloughran commented:
checkstyles and whitespace. The javac changes are all from deprecating hflush capability

mehakmeet commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCreateTest.java:
```diff
@@ -436,4 +447,110 @@ private void createFile(Path path) throws IOException {
+      try {
+        out.hflush();
+        if (!supportsFlush) {
+          // hsync not ignored
```
little confusing comment, maybe we can change this to "hysc tests not ignored"?

mehakmeet commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCreateTest.java:
```diff
@@ -436,4 +447,110 @@ private void createFile(Path path) throws IOException {
+    LOG.info("Expecting files under {} to have supportsSync={}"
+            + " and supportsFlush={}",
+        path, supportsSync, supportsFlush);
+
```
nit: blank line.

mehakmeet commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCreateTest.java:
```diff
@@ -436,4 +447,110 @@ private void createFile(Path path) throws IOException {
+          LOG.info("Successfully read synced data on a new reader {}", in);
+        }
+      } else {
+        // np sync. Let's do a flush and see what happens.
```
nit: typo in "np" -> "no"

mehakmeet commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCreateTest.java:
```diff
@@ -436,4 +447,110 @@ private void createFile(Path path) throws IOException {
+          if (!isSupported(IS_BLOBSTORE)) {
+            throw e;
+          }
+        }
```
Maybe add a LOG.warn() about the FileNotFoundException if it is an object store?

mehakmeet commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java:
```diff
@@ -137,8 +139,13 @@ public void initialize(URI uri, Configuration conf) throws IOException {
+
     public LocalFSFileInputStream(Path f) throws IOException {
       fis = new FileInputStream(pathToFile(f));
+      bytesRead = ioStatistics.getCounterReference(
```
No issues here, just confirming if these changes aren't part of a different PR by mistake(IOStats PR).

steveloughran replied:
this is a followon to IOStats...it was actually blocked on the buffer between row local fs output and FSDataOutput:
`org.apache.hadoop.fs.statistics.BufferedIOStatisticsOutputStream` - needed for hsync passthrough to the File output stream. We're now generating IOStats for the local FS too.

mehakmeet reviewed (commented):
+1, pending by some nits. I have already reviewed the .md in the previous pr(now closed) and was satisfied.

mehakmeet commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/localfs/TestLocalFSContractCreate.java:
```diff
@@ -18,7 +18,11 @@
+import org.junit.Test;
+
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
```
unused import.

steveloughran commented:
FWIW, here are the IOStats logged on the runs

ABFS (which supports hsync)
```
.dfs.core.windows.net/stevel-testing/test/testSyncable?action=flush&retainUncommittedData=false&position=2&close=true&timeout=90
2021-02-08 12:27:48,642 INFO  [JUnit-testSyncable]: contract.AbstractContractCreateTest (AbstractContractCreateTest.java:validateSyncableSemantics(586)) - IOStatistics counters=((queue_shrunk_ops=1) (time_spent_task_wait.failures=0) (write_current_buffer_ops=2) (bytes_upload_failed=0) (bytes_upload=2) (time_spent_on_put_request.failures=0) (time_spent_on_put_request=2) (time_spent_task_wait=0) (bytes_upload_successfully=2));
gauges=();
minimums=((time_spent_task_wait.failures.min=-1) (time_spent_on_put_request.min=28) (time_spent_task_wait.min=-1) (time_spent_on_put_request.failures.min=-1));
maximums=((time_spent_on_put_request.failures.max=-1) (time_spent_task_wait.max=-1) (time_spent_on_put_request.max=36) (time_spent_task_wait.failures.max=-1));
means=((time_spent_task_wait.mean=(samples=0, sum=0, mean=0.0000)) (time_spent_on_put_request.failures.mean=(samples=0, sum=0, mean=0.0000)) (time_spent_task_wait.failures.mean=(samples=0, sum=0, mean=0.0000)) (time_spent_on_put_request.mean=(samples=2, sum=64, mean=32.0000)));
```

s3a

```
2021-02-08 12:29:05,294 [JUnit-testSyncable] WARN  contract.AbstractContractCreateTest (AbstractContractCreateTest.java:validateSyncableSemantics(575)) - Output file was not created; this is an object store with different visibility semantics
2021-02-08 12:29:05,500 [JUnit-testSyncable] INFO  contract.AbstractContractCreateTest (AbstractContractCreateTest.java:validateSyncableSemantics(586)) - IOStatistics counters=((stream_write_bytes=2) (stream_write_queue_duration=0) (action_executor_acquired=1) (stream_write_block_uploads=1) (stream_write_exceptions=0) (stream_write_exceptions_completing_upload=0) (stream_write_total_time=0) (stream_write_total_data=2) (action_executor_acquired.failures=0));
gauges=((stream_write_block_uploads_data_pending=0) (stream_write_block_uploads_pending=1));
minimums=((action_executor_acquired.failures.min=-1) (action_executor_acquired.min=0));
maximums=((action_executor_acquired.max=0) (action_executor_acquired.failures.max=-1));
means=((action_executor_acquired.failures.mean=(samples=0, sum=0, mean=0.0000)) (action_executor_acquired.mean=(samples=1, sum=0, mean=0.0000)));
```

steveloughran committed:
HADOOP-13327: mehakmeets comments

Plus
* validate getFileStatus() behavior, allow for FS to not immediately update
  it
* declare that stream impls MUST implement and then reject syncable so
  that apps calling it don't get mislead.

Change-Id: Id53f48d53652963e6d9f8f6e322a524c89106b27

mukund-thakur commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StoreImplementationUtils.java:
```diff
@@ -0,0 +1,96 @@
+   * Probe for an input stream having a capability; returns true
+   * if the stream implements {@link StreamCapabilities} and its
+   * {@code hasCapabilities()} method returns true for the capability.
+   * @param out output stream
```
input stream

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+## Output Stream Model
+
+For this specification, an output stream can be viewed as a list of bytes
+stored in the client -the `hsync()` and `hflush()` operations the actions
```
are the actions?

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,1002 @@
+FS'.Files(path) == buffer
+```
+
+Any client reading the data at the path MUST see the new data.
````
As per the java doc on hsync and hflush,they are about the guarantees on data being actually writted to disk or final storage. So just a question here - How it is made sure that data will be visible after close call for the readers?

steveloughran replied:
up to the implementation. Close() must pass all its data to the shared FS.

Now, if you want some fun, look at [NFS client side caching](https://docstore.mik.ua/orelly/networking_2ndEd/nfs/ch07_04.htm). Dates from the era of diskless sun workstations and was optimised for short lived files which would only be used by the workstations, so copying over a 1MB/s shared ethernet to an even slower shared HDD would hurt the rest of the cluster.

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+`FSDataOutputStreamBuilder.build()` calls return an instance
+of a class `FSDataOutputStream`, a subclass of `java.io.OutputStream`.
+
+The base class wraps an `OutputStream` instance, one which may implement `Streamable`,
```
Syncable not Streamable right?

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+
+Forwarding this to a full flush across a distributed filesystem, or worse,
+a distant object store, is very inefficient.
+Filesystem clients which do uprate a `flush()` to an `hflush()` will eventually
```
update

steveloughran replied:
I went with "upgrade" in the end

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+Follow-on calls to `close()` are ignored, and calls to other methods
+rejected. That is: caller's cannot be expected to call `close()` repeatedly
+until it succeeds.
+1. The duration of the `call()` operation is undefined. Operations which rely
```
close() not call()?

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+it MAY be durable. That is: it does not have to be persisted, merely guaranteed
+to be consistently visible to all clients attempting to open a new stream reading
+data at the path.
+1. `Syncable.hsync()` MUST transmit the data as per `hflush` the data and persist
```
remove the second "the data" ?

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+to is not something everyone expects, nor convenient for any program trying
+to pick up updated data in a file being written. Most visible is the length
+of a file returned in the various `list` commands and `getFileStatus` —this
+is often out of data.
```
data or date?

steveloughran replied:
date. will fix.

mukund-thakur commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,1002 @@
+By default, HDFS does not sync data to disk when a stream is closed; it will
+be asynchronously saved to disk.
+
+This does not mean that users do not expect it.
```
This one is the best :)

mukund-thakur reviewed (approved):
LGTM +1,  pending some nits.
Awesome work on the docs. @steve

steveloughran committed:
HADOOP-13327. Mukunds comments.

if all is good on  yetus, will merge

Change-Id: I23344abdf040649985c12222fef71e57c1ae7fcf

steveloughran commented:
done the final comments and then rebased; unless yetus really blows up will merge in

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 40s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |   |   0m  0s | [test4tests](test4tests) |  The patch appears to include 11 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m  1s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  26m 18s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  27m 42s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |  23m 17s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  checkstyle  |   4m 51s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   7m  8s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  30m  6s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   5m  4s |  |  trunk passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   6m 11s |  |  trunk passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +0 :ok: |  spotbugs  |   0m 53s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |  11m 36s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 48s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m 24s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 16s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| -1 :x: |  javac  |  20m 16s | [/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/artifact/out/diff-compile-javac-root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04.txt) |  root-jdkUbuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 generated 11 new + 2026 unchanged - 0 fixed = 2037 total (was 2026)  |
| +1 :green_heart: |  compile  |  18m  2s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| -1 :x: |  javac  |  18m  2s | [/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/artifact/out/diff-compile-javac-root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01.txt) |  root-jdkPrivateBuild-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 generated 11 new + 1919 unchanged - 0 fixed = 1930 total (was 1919)  |
| -0 :warning: |  checkstyle  |   3m 55s | [/diff-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/artifact/out/diff-checkstyle-root.txt) |  root: The patch generated 1 new + 183 unchanged - 7 fixed = 184 total (was 190)  |
| +1 :green_heart: |  mvnsite  |   6m 37s |  |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s | [/whitespace-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/artifact/out/whitespace-eol.txt) |  The patch has 2 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  xml  |   0m  6s |  |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  shadedclient  |  13m 16s |  |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   5m  7s |  |  the patch passed with JDK Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   6m 12s |  |  the patch passed with JDK Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01  |
| +1 :green_heart: |  findbugs  |  12m 13s |  |  the patch passed  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  17m 16s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 37s |  |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  | 191m  0s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 16s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 23s |  |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 22s |  |  hadoop-azure-datalake in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  9s |  |  The patch does not generate ASF License warnings.  |
|  |   | 460m 25s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2587 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle markdownlint xml |
| uname | Linux 8d9eac62d9ae 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 462561654bb |
| Default Java | Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9.1+1-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_275-8u275-b01-0ubuntu1~20.04-b01 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/testReport/ |
| Max. process+thread count | 3757 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-aws hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2587/4/console |
| versions | git=2.25.1 maven=3.6.3 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

ndimiduk commented:
Thank you for this effort, @steveloughran !

steveloughran commented:
@ndimiduk this will let you run HBase against local FS now as if you turn off the checksumming the WAL will be synced properly...

ndimiduk commented:
No one should do this. Only want it for stable tests.

steveloughran commented:
> No one should do this. Only want it for stable tests.

that and to challenge sqllite for dominance in the embedded-database market
